Deploy production ordering service doc (#2080)
hyperledger/fabric: docs/source/deployorderer/ordererdeploy.md

Finally, your ordering service includes a consortium of members as defined in the `Organization` section of the channel configuration. Additional application channels can be created for custom combinations of ordering service consortium members to transact privately by using the `cfgtxgen` command and the `configtx.yaml` file to generate the genesis block for the new channel.

## Troubleshooting

### When you start the orderer, it fails with the following error:
```
ERRO 001 failed to parse config:  Error reading configuration: Unsupported Config Type ""
```

**Solution:**
Your `FABRIC_CFG_PATH` is not set. Run the following command to set it to the location of your `orderer.yaml` file.

```
export FABRIC_CFG_PATH=<PATH_TO_ORDERER_YAML>
```

### When you start the orderer, it fails with the following error:
```
PANI 003 Failed to setup local msp with config: administrators must be declared when no admin ou classification is set
```

**Solution:**   

> Your local MSP definition is missing the `config.yaml` file. Create the file and copy it into the local MSP `/msp` folder of orderer. See the [Fabric CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous) documentation for more instructions.

Reviewer: I think I noted this in the channel participation tutorial as well but which `config.yaml` are we referring to here?

Author: `config.yaml` is a file that enables Node OUs when it's included in an MSP.

### When you start the orderer, it fails with the following error:
```
PANI 005 Failed validating bootstrap block: initializing channelconfig failed: could not create channel Orderer sub-group config: setting up the MSP manager failed: administrators must be declared when no admin ou classification is set
```

**Solution:**   

> The system channel configuration is missing `config.yaml` file. If you are creating a new ordering service, the `MSPDir` referenced in `configtx.yaml` file is missing the `config.yaml` file. Follow instructions in the [Fabric CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous) documentation to generate this file and then rerun `configtxgen` to regenerate the genesis block for the system channel.

Reviewer: Same `config.yaml` comment here.

- When you enroll the orderer identity with the TLS CA, the public key is generated in the `signcerts` folder, and the private key is located in the `keystore` directory. Rename the private key in the `keystore` folder to `orderer0-tls-key.pem` so that it can be easily recognized later as the TLS private key for this node.
- Copy the orderer TLS certificate and private key files to `organizations/ordererOrganizations/ordererOrg1.example.com/orderers/orderer0.ordererOrg1.example.com/tls`. The path and name of the certificate and private key files correspond to the values of the `General.TLS.Certificate` and `General.TLS.PrivateKey` parameters in the `orderer.yaml`.
- If you require mutual TLS (`General.TLS.ClientAuthRequired` set to true), you need to indicate to the orderer which TLS CA root certificates to use to authorize clients. Copy the organization's TLS CA root certificate `ca-cert.pem` to `organizations/ordererOrganizations/ordererOrg1.example.com/orderers/orderer0.ordererOrg1.example.com/tls/tls-cert.pem` so that the organization's clients will be authorized.  If `General.TLS.ClientAuthRequired` is set to false, you can skip this step.

> **Note:** Don't forget to create the `config.yaml` file and add it to the organization MSP and local MSP folder for each ordering node. This file enables Node OU support for the channel, an important feature that allows the channel to recognize certificates that contain the "admin" OU as an admin identity of the channel. Learn more in the [Fabric CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous) documentation.

Reviewer 2: This paragraph wasn't quite right... I believe it was also copy/pasted into the channel participation tutorial, and perhaps elsewhere.

```suggestion
**Note:** Don't forget to create the `config.yaml` file and add it to the organization MSP and local MSP folder for each ordering node. This file enables Node OU support for the MSP, an important feature that allows the MSP's admin to be identified based on an "admin" OU in an identity's certificate. Learn more in the [Fabric CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html#nodeous) documentation.
```

Reviewer 3: Done.

> If you are using a containerized solution for running your network (which for obvious reasons is a popular choice), **it is a best practice to mount these folders (volumes) external to the container where the node itself is running. This will allow the certificates to be used to create a new node should the node container go down, become corrupted, or is restarted.**

Reviewer 2: ```suggestion
If you are using a containerized solution for running your network (which for obvious reasons is a popular choice), **it is a best practice to mount volumes for the certificate directories external to the container where the node itself is running. This will allow the certificates to be used by an ordering node container, regardless whether the ordering node container goes down, becomes corrupted, or is restarted.**
```

Reviewer 3: Done.

#### Set up the `configtxgen` tool

All channels are created by building a channel creation transaction and submitting the transaction to the ordering service. The channel creation transaction specifies the initial configuration of the channel and is used by the ordering service to write the channel genesis block. While it is possible to build the channel creation transaction file manually, it is easier to use the [configtxgen](../commands/configtxgen.html) tool, which works by reading a `configtx.yaml` file that defines the configuration of your channel and then writing the relevant information into a configuration block known as the "genesis block".

Notice that the `configtxgen` tool is located in the `bin` folder of downloaded Fabric binaries.

Before using `configtxgen`, confirm you have set the `FABRIC_CFG_PATH` environment variable to the path of the directory that contains your local copy of the `configtx.yaml` file. You can verify that are able to use the tool by printing the `configtxgen` help text:

```
configtxgen --help
```

#### The `configtx.yaml` file

> The `configtx.yaml` file is used to specify the **channel configuration** of new channels. The information that is required to build the channel configuration is specified in a readable and editable form in the `configtx.yaml` file. The `configtxgen` tool uses the channel profiles that are defined in `configtx.yaml` to create the channel configuration block in the [protobuf format](https://developers.google.com/protocol-buffers).

Reviewer 2: ```suggestion
The `configtx.yaml` file is used to specify the **channel configuration** of the system channel and application channels. The information that is required to build the channel configuration is specified in a readable and editable form in the `configtx.yaml` file. The `configtxgen` tool uses the channel profiles that are defined in `configtx.yaml` to create the channel configuration block in the [protobuf format](https://developers.google.com/protocol-buffers).
```

Reviewer 3: Done.

- **Profiles** Each channel profile references information from other sections of the `configtx.yaml` file to build a channel configuration. The profiles are used to create the genesis block of the channel.

The `configtxgen` tool uses `configtx.yaml` file to create the genesis block for the channel. A detailed version of the `configtx.yaml` file is available in the [Fabric sample configuration](https://github.com/hyperledger/fabric/blob/{BRANCH}/sampleconfig/configtx.yaml). Refer to the [Using configtx.yaml to build a channel configuration](../create_channel/create_channel_config.html) tutorial to learn more about the  settings in this file.

#### Generate the genesis block

After you have completed editing the `configtx.yaml`, you can start the process of creating a system channel for the ordering service. Every channel configuration starts with a genesis block. We have already set the environment variables that are required to use the `configtxgen` tool. Run the following command to build the genesis block for `channel1` using the `SampleAppGenesisEtcdRaft` profile:

```

> configtxgen -profile SampleAppGenesisEtcdRaft -outputCreateChannelTx ./channel-artifacts/channel1.tx -channelID channel1

Reviewer 2: Hang on, the first sentence of this section said we were going to create the genesis block of the system channel. We've jumped straight to an application channel creation here.

We need to do the steps in this order:
1) generate the system channel genesis block
2) use the system channel genesis block to bootstrap ordering nodes
3) generate application channel create transaction
4) create the application channel on the ordering service

The 'create a channel' tutorial already does these steps (at least steps 1,3,4). Do we want to document the steps twice, or just point to the create a channel tutorial?

Reviewer 3: Good catch.
I grabbed the wrong profile. 
I've updated the instructions to be more specific to the system channel genesis block. 
and used the `TwoOrgsOrdererGenesis:` profile  from the Test network instead.

For steps 3 and 4 I link to those instructions in the create a channel tutorial.

```

You should have already used your certificate authority of choice to generate the orderer enrollment certificate, TLS certificate, private keys, and the MSPs that Fabric must consume. Refer to the [CA deployment guide](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/cadeploy.html) and [Registering and enrolling identities with a CA](https://hyperledger-fabric-ca.readthedocs.io/en/release-1.4/deployguide/use_CA.html) topics for instructions on how to create a Fabric CA and how to generate these certificates. You need to generate the following sets of certificates:
  - **Orderer organization MSP**
  - **Orderer TLS CA certificates**
  - **Orderer local MSP (enrollment certificate and private key of the orderer)**

You will either need to use the Fabric CA client to generate the certificates directly into the recommended folder structure or you will need to copy the generated certificates to their recommended folders after they are generated. Whichever method you choose, most users are ultimately likely to script this process so it can be repeated as needed. A list of the certificates and their locations is provided here for your convenience.

> If you are using a containerized solution for running your network (which for obvious reasons is a popular choice), **it is a best practice to mount these folders (volumes) external to the container where the node itself is running. This will allow the certificates to be used to create a new node should the node container go down, become corrupted, or is restarted.**

Reviewer 2: My prior comment on this paragraph was not addressed, even though Pam said "Done". Makes me wonder if a whole set of updates was missed in the latest commit?

https://github.com/hyperledger/fabric/pull/2080#discussion_r523108405

#### TLS certificates

For the ordering node to launch successfully, the locations of the TLS certificates you specified in the [Checklist for a production ordering node](./ordererchecklist.html) must point to the correct certificates. To do this:

> - Copy the **TLS CA Root certificate** that contains the public key that is associated with the signing (private) key certificate, which by default is called `ca-cert.pem`, to the orderer organization MSP definition  `organizations/ordererOrganizations/ordererOrg1.example.com/msp/tlscacerts/tls-cert.pem`.

Reviewer 2: Was this a copy/paste from somewhere? The description is confusing... I was expecting it to describe what a TLS CA root certificate is, not what a certificate is in general.

Author: It's basically identical to what we have in the peer deployment guide.

Reviewer 2: Ok, that's why I asked if it was copy/paste, so that we can fix in both places.  Lets just leave it as "TLS CA Root certificate", I feel talking about public and private signing keys are only going to confuse people, as it gives the impression these certs are going to be used for signing transactions or communications. They're not, they're only used for verifying issued certs.

Author: Ok done.

> - Copy the **CA Root certificate** that contains the public key that is associated with the signing (private) key certificate, which by default is called `ca-cert.pem`, to the orderer organization MSP definition  `organizations/ordererOrganizations/ordererOrg1.example.com/msp/cacerts/ca-cert.pem`.

Reviewer 2: Was this a copy/paste from somewhere? The description is confusing... I was expecting it to describe what a CA root certificate is, not what a certificate is in general.

Author: It's basically identical to what we have in the peer deployment guide.

Reviewer 2: Ok, that's why I asked if it was copy/paste, so that we can fix in both places.  Lets just leave it as "CA Root certificate", I feel talking about public and private signing keys are only going to confuse people, as it gives the impression these certs are going to be used for signing transactions or communications. They're not, they're only used for verifying issued certs.

Author: Ok done.

- When you enroll the orderer identity with the TLS CA, the public key is generated in the `signcerts` folder, and the private key is located in the `keystore` directory. Rename the private key in the `keystore` folder to `orderer0-tls-key.pem` so that it can be easily recognized later as the TLS private key for this node.
- Copy the orderer TLS certificate and private key files to `organizations/ordererOrganizations/ordererOrg1.example.com/orderers/orderer0.ordererOrg1.example.com/tls`. The path and name of the certificate and private key files correspond to the values of the `General.TLS.Certificate` and `General.TLS.PrivateKey` parameters in the `orderer.yaml`.

> - If you require mutual TLS (`General.TLS.ClientAuthRequired` set to true), you need to indicate to the orderer which TLS CA root certificates to use to authorize clients. Copy the TLS CA root certificate `ca-cert.pem` to `organizations/ordererOrganizations/ordererOrg1.example.com/orderers/orderer0.ordererOrg1.example.com/tls/tls-cert.pem` so that the organization's clients will be authorized. If more than one organization is going to connect to this organization, you would need to add their TLS certificates also. If `General.TLS.ClientAuthRequired` is set to false, you can skip this step.

Reviewer 2: I believe this entire bullet (line 80) relates to General.TLS.ClientRootCAs, which is not common to be set, as explained in the property checklist. Therefore I'd recommend to remove this entire bullet, as it is more likely to confuse than help. In the rare cases where people need to set it, they can see the reference explanation in the checklist.

> - If you require mutual TLS (`General.TLS.ClientAuthRequired` set to true), you need to indicate to the orderer which TLS CA root certificates to use to authorize clients. Copy the organization's TLS CA root certificate `ca-cert.pem` to `organizations/ordererOrganizations/ordererOrg1.example.com/orderers/orderer0.ordererOrg1.example.com/tls/tls-cert.pem` so that the organization's clients will be authorized.  If `General.TLS.ClientAuthRequired` is set to false, you can skip this step.

Reviewer 2: The TLS CA root cert from EACH org that is going to connect to the orderer should be copied, that is, application clients and peers that will connect to this orderer.

Author: In this tutorial, there's only one org. But I'll add a clarification note here anyway.

Note that the local MSP contains the signed certificate (public key) and the private key for the orderer. The private key is used by the node to sign transactions, and is therefore not shared and must be secured. For maximum security, a Hardware Security Module (HSM) can be configured to generate and store this private key.

### Create the ordering service genesis block

> The first channel that is created in a Fabric network is the "system" channel. The system channel defines the set of ordering nodes that form the ordering service and the set of organizations that serve as ordering service administrators. Peers transact on private "application" channels that are derived from the ordering service system channel, which also defines the "consortium" (the peer organizations known to the ordering service). Therefore, before you can deploy an ordering service, you need to generate the system channel configuration by creating the "genesis block". We use the system channel genesis block to bootstrap each ordering node. Fabric provides the `configtxgen` tool for this purpose.

Reviewer 2: ```suggestion
The first channel that is created in a Fabric network is the "system" channel. The system channel defines the set of ordering nodes that form the ordering service and the set of organizations that serve as ordering service administrators. Peers transact on private "application" channels that are derived from the ordering service system channel, which also defines the "consortium" (the peer organizations known to the ordering service). Therefore, before you can deploy an ordering service, you need to generate the system channel configuration by creating the "genesis block". Fabric provides the `configtxgen` tool for this purpose. We'll then use the generated system channel genesis block to bootstrap each ordering node. 
```

> The first channel that is created in a Fabric network is the "system" channel. The system channel defines the set of ordering nodes that form the ordering service and the set of organizations that serve as ordering service administrators. Peers transact on private "application" channels that are derived from the ordering service system channel, which also defines the "consortium" (the peer organizations known to the ordering service). Therefore, before you can deploy an ordering service, you need to generate the system channel configuration by creating the "genesis block". abric provides the `configtxgen` tool for this purpose. We'll then use the generated system channel genesis block to bootstrap each ordering node.

Reviewer 2: ```suggestion
The first channel that is created in a Fabric network is the "system" channel. The system channel defines the set of ordering nodes that form the ordering service and the set of organizations that serve as ordering service administrators. Peers transact on private "application" channels that are derived from the ordering service system channel, which also defines the "consortium" (the peer organizations known to the ordering service). Therefore, before you can deploy an ordering service, you need to generate the system channel configuration by creating the system channel "genesis block". Fabric provides the `configtxgen` tool for this purpose. We'll then use the generated system channel genesis block to bootstrap each ordering node.
```

#### Set up the `configtxgen` tool

> All channels are created by building a channel creation transaction and submitting the transaction to the ordering service. The channel creation transaction specifies the initial configuration of the channel and is used by the ordering service to write the channel genesis block. While it is possible to build the channel creation transaction file manually, it is easier to use the [configtxgen](../commands/configtxgen.html) tool, which works by reading a `configtx.yaml` file that defines the configuration of your channel and then writing the relevant information into a configuration block known as the "genesis block".

Reviewer 2: This paragraph seems to talk about application channels, while the prior paragraph and next paragraphs talk about system channel genesis block creation. I'd highlight that configtxgen is used for both.

Author: I don't get an "application channel" vibe from this. Not sure exactly what you're referring to. I changed this first sentence to: 

"Both the system channel and all application channels are created by building a channel creation transaction and submitting the transaction to the ordering service."

Does that help?

Reviewer 2: No. 
"All channels are created by building a channel creation transaction and submitting the transaction to the ordering service."
That is true for application channels.
It is not true for system channels. First of all, if you are creating the system channel there is not yet an ordering service to submit to!  Second of all, for the system channel you generate a system channel genesis block (not a channel create transaction).

Author: The simplest solution here seemed to be to just cut everything before the sentence beginning with "While it is possible", as that's really what's relevant to setting up configtxgen.

- **Orderer:** Which ordering nodes will form the Raft consenter set of the channel.
- **Policies** Different sections of the file work together to define the channel policies that will govern how organizations interact with the channel and which organizations need to approve channel updates. For the purposes of this tutorial, we will use the defaults that are used by Fabric. For more information about policies, check out [Policies](../policies/policies.html).
- **Profiles** Each channel profile references information from other sections of the `configtx.yaml` file to build a channel configuration. The profiles are used to create the genesis block of the channel.

The `configtxgen` tool uses `configtx.yaml` file to create the genesis block for the channel. A detailed version of the `configtx.yaml` file is available in the [Fabric sample configuration](https://github.com/hyperledger/fabric/blob/{BRANCH}/sampleconfig/configtx.yaml). Refer to the [Using configtx.yaml to build a channel configuration](../create_channel/create_channel_config.html) tutorial to learn more about the  settings in this file.

> #### Generate the genesis block

Reviewer 2: ```suggestion
#### Generate the system channel genesis block
```

After you have completed editing the `configtx.yaml` to reflect the orderer and peer organizations that will participate in your network, run the following command to create the genesis block of the system channel:
```
configtxgen -profile TwoOrgsOrdererGenesis -channelID system-channel -outputBlock ./system-genesis-block/genesis.block
```

Where:
- `-profile` refers to the `TwoOrgsOrdererGenesis` profile in `configtx.yaml`.

> - `-channelID` must be `system-channel`.

Reviewer 2: Doesn't have to be "system-channel", can be any name.

> - `-outputCreateChannelTx` refers to the location of the generated genesis block.

Reviewer 2: ```suggestion
- `outputBlock` refers to the location of the generated genesis block.
```

When the command is successful, you will see logs of `configtxgen` loading the `configtx.yaml` file and printing a channel creation transaction:
```
INFO 001 Loading configuration
INFO 002 Loaded configuration: /Usrs/fabric-samples/test-network/configtx/configtx.yaml

> INFO 003 Generating new channel configtx
> INFO 004 Writing new channel tx

Reviewer 2: ```suggestion
INFO 004 Generating genesis block
INFO 005 Creating system channel genesis block
INFO 006 Writing genesis block
```

```

> Make note of the generated `.pb` file. This is your genesis block and will be referenced in the `orderer.yaml` file below.

Reviewer 2: ```suggestion
Make note of the generated output block filename. This is your genesis block and will be referenced in the `orderer.yaml` file below.
```

### Configuration of `orderer.yaml`

Now you can use the [Checklist for a production orderer](./ordererchecklist.html) to modify the default settings in the `orderer.yaml` file. In the future, if you decide to deploy the orderer through Kubernetes or Docker, you can override the same default settings by using environment variables instead. Check out the [note](../deployment_guide_overview.html#step-five-deploy-orderers-and-ordering-nodes) in the deployment guide overview for instructions on how to construct the environment variable names for an override.

At a minimum, you need to configure the following parameters:
- `General.ListenAddress` - Hostname that the ordering node listens on.
- `General.ListenPort` - Port that the ordering node listens on.
- `General.TLS.Enabled: true` - Server-side TLS should be enabled in all production networks.
- `General.TLS.PrivateKey ` - Ordering node private key from TLS CA.
- `General.TLS.Certificate ` - Ordering node signed certificate (public key) from the TLS CA.
- `General.TLS.RootCAS` - This value should be unset.
- `General.Cluster.ClientCertificate` - Ordering node signed certificate (public key) from the TLS CA.

> - `General.Cluster.ClientPrivateKey` - Ordering node private key from TLS CA.

Reviewer 2: These two are optional as of v2.3 since they now default based on General.TLS.Certificate and General.TLS.PrivateKey

Reviewer 3: Since this will be ported back to v2.2 I think we wanted to make another pass with updates like this for v2.3 after this gets merged for 2.2

Reviewer 3: Although I did ask for this fix to be ported back to v2.2. Let me check if it made it in.

Reviewer 3: It was backported [here](https://github.com/hyperledger/fabric/pull/2078#issuecomment-726768203) so we can add it

Author: Which too? Was this added and I missed it? @pamandrejko

Reviewer 3: We have covered this in the checklist where we say:
As of Fabric v2.3.0, if unset, the `ClientCertificate` and `ClientPrivateKey` default to the server `General.TLS.Certificate` and `General.TLS.PrivateKey` when the orderer is not configured to use a separate cluster port.

Therefore I think we can remove the following two parameters from here since it would not be a typical use case to specify them.
- `General.Cluster.ClientCertificate` - Ordering node signed certificate (public key) from the TLS CA.
- `General.Cluster.ClientPrivateKey` - Ordering node private key from TLS CA.

- `General.BoostrapMethod:file` - Start ordering service with a system channel.

> - `General.BootstrapFile` - Path to and name of the genesis block `.pb` file for the ordering service system channel.

Reviewer 2: ```suggestion
- `General.BootstrapFile` - Path to and name of the genesis block file for the ordering service system channel.
```

    ```
    export FABRIC_CFG_PATH=../config
    ```

After `orderer.yaml` has been configured and your deployment backend is ready, you can simply start the orderer node with the following command:

```
cd bin
./orderer start
```

When the orderer starts successfully, you should see a message similar to:

```
INFO 019 Starting orderer:
INFO 01a Beginning to serve requests
```

You have successfully started one node, you now need to repeat these steps to configure and start the other two orderers. When a majority of orderers are started, a Raft leader is elected. You should see something similar to the following output:
```
INFO 01b Applied config change to add node 1, current nodes in channel: [1] channel=syschannel node=1
INFO 01c Applied config change to add node 2, current nodes in channel: [1 2] channel=syschannel node=1
INFO 01d Applied config change to add node 3, current nodes in channel: [1 2 3] channel=syschannel node=1
INFO 01e raft.node: 1 elected leader 2 at term 11 channel=syschannel node=1
INFO 01f Raft leader changed: 0 -> 2 channel=syschannel node=1
```

## Next steps

> Your ordering service is started and ready to order transactions into blocks. Working together to form the ordering service, the orderers receive transactions from application clients and package the transactions into a well-defined sequence of blocks. These blocks form the blockchain and are ultimately distributed to all peers on the channel to be committed to their ledger.

Reviewer 2: Actually, at this point, we are not ready to receive transactions from application clients, since we've only created the system channel, we haven't crated application channels yet.
